### PR TITLE
Parameter setting cmm algo

### LINKED ISSUE #4984: Parameter setting of the CMM calculation algorithm to be used
There are several algorithms for the calculation of the average monthly consumption, and with the BHIMA system we had developed 4 algorithms, this issue proposes to be able to define in stock setting the algorithm to use

### PR DESCRIPTION
**Pull request by lomamech**
- Add property average consumption algorith is stock setting
- Update form for manage stock setting
- Add define algorith to use for average monthly consuption to Use

closes #4984

### CODE REVIEW

**Comment by lomamech** [comment]
@jniles @mbayopanda  @jmcameron  All checks have passed

File: client/src/i18n/en/stock.json
Comment (mbayopanda) [inline client/src/i18n/en/stock.json:106]: Since you mentioned `monthly consumption` and `30.5` which are related to month, is it possible to change `period defined` by just `month`?
Ex. The average monthly consumption is obtained by dividing the quantity consumed during the month by the number of days of stock of the month, and by multiplying the result obtained by 30.5

**Review by mbayopanda — APPROVED** [review]
LGTM 👍

**Comment by jniles** [comment]
@lomamech looks like this will need to have some merge fixes before it can be merged.

File: client/src/modules/stock/settings/stock-settings.html
Comment (jmcameron) [inline client/src/modules/stock/settings/stock-settings.html:52]: Great idea adding these tags so we can access them easily in the tests!

File: client/src/modules/stock/settings/stock-settings.js
Comment (jmcameron) [inline client/src/modules/stock/settings/stock-settings.js]: I find it hard to keep track of which algorithm is 1, 2, etc.   I suggest appending the algorithm number on these functions:    getAvertageConsomptionAlgo1()     (add the correct number!)
That way is is easier to track what is happening.

File: test/end-to-end/stock/stock.setting.js
Comment (jmcameron) [inline test/end-to-end/stock/stock.setting.js:26]: Can we verify that these items have actually changed?  (also in the next 'it').

File: server/models/schema.sql
Comment (jmcameron) [inline server/models/schema.sql]: It is not clear to me why we need a table for the algorithms?   They seem to be something that is strictly in the Javascript code.

**Review by jmcameron — APPROVED** [review]
Great job!   

I ran this on system ('yarn dev') and it seems to work fine.

The new tests ran fine ('yarn test:ends').
Note that the unrelated distribution keys failed on my system (see Issue https://github.com/IMA-WorldHealth/bhima/issues/5006).  Since it passes on the CI testing, this is nothing for you to worry about.

I'm unclear: I see that we can select the CMM algorithm, etc, but are we actually using that algorithm in the CMM calculations (eg on the Items in Stock page)?

Also, when we start using the selected algorithm in the items in stock page, I suggest adding a tool tip on the "CMM" column header that shows which algorithm is being used for the calculation (when you hover over it).

File: client/src/i18n/en/stock.json
Comment (lomamech) [reply client/src/i18n/en/stock.json:106]: I had agreed with @mbayopanda, for the calculation of the CMM we do not use only one me but rather the period defined in the stock Setting 
```
month_average_consumption 
````

**Comment by lomamech** [comment]
> Great job!
> 
> I ran this on system ('yarn dev') and it seems to work fine.
> 
> The new tests ran fine ('yarn test:ends').
> Note that the unrelated distribution keys failed on my system (see Issue #5006). Since it passes on the CI testing, this is nothing for you to worry about.
> 
> I'm unclear: I see that we can select the CMM algorithm, etc, but are we actually using that algorithm in the CMM calculations (eg on the Items in Stock page)?
> 
> Also, when we start using the selected algorithm in the items in stock page, I suggest adding a tool tip on the "CMM" column header that shows which algorithm is being used for the calculation (when you hover over it).

We are not using it yet, and it will be dealt with in the following issue, 

https://github.com/IMA-WorldHealth/bhima/issues/4976

File: client/src/modules/stock/settings/stock-settings.js
Comment (lomamech) [reply client/src/modules/stock/settings/stock-settings.js]: This function just allows you to obtain the different existing algorithms...
````
// GET '/stock/setting/average_consuption'
// Get the different algorithms for calculating the average monthly consumption
exports.averageConsuptionAlgo = function averageConsuptionAlgo(req, res, next) {
  const sql = `SELECT id, name, label, comment FROM average_consumption_algo;`;

  db.exec(sql)
    .then((rows) => {
      res.status(200).json(rows);
    })
    .catch(next)
    .done();
};

````
image

File: server/models/schema.sql
Comment (lomamech) [reply server/models/schema.sql]: The reason we used a table is just because we had thought to keep the identifier of the algorithm to be used among the stock settings.

**Comment by lomamech** [comment]
@jniles @mbayopanda @mbayopanda this PR need another review or Approbation

File: client/src/modules/stock/settings/stock-settings.js
Comment (jmcameron) [reply client/src/modules/stock/settings/stock-settings.js]: Ok.  Thx.

File: server/models/schema.sql
Comment (jmcameron) [reply server/models/schema.sql]: That sounds fine.  Can't we use an enum for the algorithm and add a new field in the stock_settings table?  I'm still not clear why we need a new table for the algorithm info.

Comment (lomamech) [reply server/models/schema.sql]: In this case we could use the component **bhConstants** and modify the property
````
 `average_consumption_algo_id` TINYINT(3) UNSIGNED NOT NULL DEFAULT 3,
````
In VARCHAR with the following values 'algo1', 'algo2', 'algo' and 'algo_msh'

Comment (lomamech) [reply server/models/schema.sql]: I will make this modification

**Comment by lomamech** [comment]
@jmcameron @jniles @mbayopanda  All Checks have passed

File: client/src/i18n/en/stock.json
Comment (jniles) [reply client/src/i18n/en/stock.json:106]: We should probably come up with a better name than "algo 1", "algo 2", etc.  Also, I'm not sure we should support anything other than algo 1 and 4.

File: client/src/i18n/fr/stock.json
Comment (jniles) [inline client/src/i18n/fr/stock.json:195]: :+1:

File: server/models/migrations/next/migrate.sql
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: Awesome 👍  Thanks for adding these to the migration script.

File: test/end-to-end/stock/stock.setting.js
Comment (jniles) [reply test/end-to-end/stock/stock.setting.js:26]: Looks like this was addressed.

**Review by jniles — APPROVED** [review]
@lomamech great work!  This is a nice and concise PR.

I agree with @jmcameron that we don't really need to store these in the database.  However, I imagine on the long term, we may end up moving a lot of our constants into the database.  You've done a good implementation, if it becomes obvious that they should or should not be in the database, we'll take a further action.

bors r+

File: server/models/migrations/next/migrate.sql
Comment (lomamech) [reply server/models/migrations/next/migrate.sql]: Sorry, I must update the migrate filte

**Pull request merged**
